### ISSUE TITLE
NaNs in prediction column in cluster mode

### ISSUE THREAD

**Issue by rmalarc** [intel-analytics/BigDL#2662]
I'm trying to create a simple 3-layer linear model (dense(25) -> dropout -> dense(1)) and get NaNs as the prediction. A single layer model works, once I add more than one layer I get the NaNs. Here's my model definition:

```
import com.intel.analytics.bigdl.dlframes.DLEstimator
import com.intel.analytics.bigdl.utils.Shape

import com.intel.analytics.bigdl.tensor.Tensor

import com.intel.analytics.bigdl.nn.MSECriterion
import com.intel.analytics.bigdl.nn.keras._
import com.intel.analytics.bigdl.optim.LBFGS
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric.NumericDouble
import com.intel.analytics.bigdl.utils.Engine
import com.intel.analytics.bigdl.tensor.Tensor

import scala.collection.mutable.ArrayBuffer
import scala.util.Random


val cores = 16
val cpus = 5


Engine.init(cpus,cores,true)
val seq = (1 to 10000).map(x=>((1 to 57).map(x1=>Random.nextDouble()),Array(Random.nextDouble())))

val data = spark.sparkContext.parallelize(seq)
val df =  spark.sqlContext.createDataFrame(data).toDF("features_dense", "response")


val model = Sequential().
  add(Dense(25, inputShape=Shape(57), activation="relu")).
  add(Dropout(0.1)).
  add(Dense(1))

val criterion = MSECriterion()
val estimator = new DLEstimator(model, criterion, Array(57), Array(1)).
  setOptimMethod(new LBFGS[Double]()).
  setLearningRate(1.0).
  setBatchSize(64*cpus*cores).
  setMaxEpoch(10).
  setFeaturesCol("features_dense").
  setLabelCol("response")



val dlModel = estimator.fit(df)
dlModel.transform(df).show()

```

the output is:

```
+--------------------+-------------+----------+
|      features_dense|     response|prediction|
+--------------------+-------------+----------+
|[0.7223931, 0.027...|  [0.3623932]|     [NaN]|
|[0.1836216, 0.024...| [0.24758616]|     [NaN]|
|[0.44344407, 0.21...|  [0.9397299]|     [NaN]|
|[0.010192783, 0.3...| [0.25302747]|     [NaN]|
|[0.37726888, 0.59...| [0.18955013]|     [NaN]|
|[0.55585444, 0.84...|  [0.5171396]|     [NaN]|
|[0.67225975, 0.91...| [0.53438455]|     [NaN]|
|[0.686856, 0.5141...|[0.025672115]|     [NaN]|
|[0.3839318, 0.731...| [0.13444223]|     [NaN]|

```

The NaN prediction only happens in a cluster environment and the model has more than 1 layer. A single model layer in a cluster actually produces a valid prediction.

Thanks!

**Comment by yiheng** [comment]
@zhangxiaoli73 Please take a look at

**Comment by zhangxiaoli73** [comment]
ok, working on it

**Comment by zhangxiaoli73** [comment]
Hi, @rmalarc I run your example in cluster, but nothing wrong happened. My test env:
(1) use bigdl master or bigdl branch-0.7
(2) use spark 2.0 and spark standalone cluster
(3) spark submit command:
spark-submit --master spark://xxx:7077  --conf spark.serializer=org.apache.spark.serializer.JavaSerializer --driver-memory 40G --executor-memory 40G --executor-cores 16 --total-executor-cores 80 --class com.intel.analytics.bigdl.example.MLPipeline.UserTest xxx.jar

**Comment by zhangxiaoli73** [comment]
code as following:
`import com.intel.analytics.bigdl.dlframes.DLEstimator
import com.intel.analytics.bigdl.utils.Shape
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.nn.MSECriterion
import com.intel.analytics.bigdl.nn.keras._
import com.intel.analytics.bigdl.optim.LBFGS
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric.NumericDouble
import com.intel.analytics.bigdl.utils.Engine
import com.intel.analytics.bigdl.tensor.Tensor
import org.apache
import org.apache.spark
import org.apache.spark.SparkContext
import org.apache.spark.sql.SQLContext

import scala.collection.mutable.ArrayBuffer
import scala.util.Random


object UserTest {
  def main(args: Array[String]): Unit = {

    val cores = 16
    val cpus = 5

    Engine.init(cpus, cores, true)
    val conf = Engine.createSparkConf().setAppName("user test")
      .set("spark.task.maxFailures", "1")
    val sc = new SparkContext(conf)
    val sqlContext = SQLContext.getOrCreate(sc)

    val seq = (1 to 10000).map(x=>((1 to 57).map(x1=>Random.nextDouble()),Array(Random.nextDouble())))

    val data = sc.parallelize(seq, cores * cpus)
    val df =  sqlContext.createDataFrame(data).toDF("features_dense", "response")


    val model = Sequential().
        add(Dense(25, inputShape=Shape(57), activation="relu")).
        add(Dropout(0.1)).
        add(Dense(1))

    val criterion = MSECriterion()
    val estimator = new DLEstimator(model, criterion, Array(57), Array(1)).
      setOptimMethod(new LBFGS[Double]()).
      setLearningRate(1.0).
      setBatchSize(64*cpus*cores).
      setMaxEpoch(10).
      setFeaturesCol("features_dense").
      setLabelCol("response")


    val dlModel = estimator.fit(df)
    dlModel.transform(df).show()
  }
}`

**Comment by zhangxiaoli73** [comment]
You can try this code, and contact me for any questions.

**Comment by rmalarc** [comment]
Thanks for looking into it.

The most notorious difference is that I'm running spark 2.3.1 and bigdl from:

https://repo1.maven.org/maven2/com/intel/analytics/zoo/analytics-zoo-bigdl_0.6.0-spark_2.3.1/0.3.0-SNAPSHOT/analytics-zoo-bigdl_0.6.0-spark_2.3.1-0.3.0-SNAPSHOT.pom

using a Databricks cluster. Thoughts?  I'll test it matching you version numbers.


On Tue, Oct 23, 2018, 1:46 AM zhangxiaoli73 <notifications@github.com<mailto:notifications@github.com>> wrote:

You can try this code, and contact me for any questions.

**Comment by rmalarc** [comment]
I'm at a loss here.

I tried the 0.6 and 0.7 for spark 2.2 and 2.3.1 in Databricks with the same result. (using the deprecated engine init call Engine.init(cpus, cores, true) )

Akin to your spark-submit I initiated my cluster with:

spark.executor.cores 16 spark.cores.max 48 spark.serializer org.apache.spark.serializer.JavaSerializer

I get an exception this time while doing Engine.init


java.lang.IllegalArgumentException: requirement failed: Engine.init: Can not find spark.shuffle.blockTransferService. For details please check https://bigdl-project.github.io/master/#APIGuide/Engine/
at scala.Predef$.require(Predef.scala:224) at com.intel.analytics.bigdl.utils.Engine$.com$intel$analytics$bigdl$utils$Engine$$verify$1(Engine.scala:392) at com.intel.analytics.bigdl.utils.Engine$$anonfun$checkSparkContext$1.apply(Engine.scala:400) at com.intel.analytics.bigdl.utils.Engine$$anonfun$checkSparkContext$1.apply(Engine.scala:400) at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33) at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35) at com.intel.analytics.bigdl.utils.Engine$.checkSparkContext(Engine.scala:400) at com.intel.analytics.bigdl.utils.Engine$.init(Engine.scala:116) at line4e5e69728c7248959b7e2d0c74e5e89c25.$read$$iw$$iw$$iw$$iw$$iw$$iw.<init>(command-812839953234884:23)

Thoughts? Are you able to run the piece of code above using databricks?

Thanks!

On Tue, Oct 23, 2018 at 5:42 AM Mauricio Alarcon <rmalarc@msn.com<mailto:rmalarc@msn.com>> wrote:
Thanks for looking into it.

The most notorious difference is that I'm running spark 2.3.1 and bigdl from:

https://repo1.maven.org/maven2/com/intel/analytics/zoo/analytics-zoo-bigdl_0.6.0-spark_2.3.1/0.3.0-SNAPSHOT/analytics-zoo-bigdl_0.6.0-spark_2.3.1-0.3.0-SNAPSHOT.pom

using a Databricks cluster. Thoughts?  I'll test it matching you version numbers.


On Tue, Oct 23, 2018, 1:46 AM zhangxiaoli73 <notifications@github.com<mailto:notifications@github.com>> wrote:

You can try this code, and contact me for any questions.

**Comment by zhangxiaoli73** [comment]
@rmalarc Sorry, we just run on original spark and have no databricks cluster.
But from your error "Can not find spark.shuffle.blockTransferService", I wonder you can set spark conf in databricks spark.shuffle.blockTransferService=netty.

**Comment by zhangxiaoli73** [comment]
find some blogs about how to use bigdl on databricks: 
https://databricks.com/blog/2017/02/09/intels-bigdl-databricks.html
https://databricks-prod-cloudfront.cloud.databricks.com/public/4027ec902e239c93eaaa8714f173bcfc/5669198905533692/2626293254583012/3983381308530741/latest.html

Hope those can help you.
